File: B트랙/8주차/참고 코드/dijkstra.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;

typedef struct edge
{
	int vertex;
	ll value;
}edge;

struct cmp
{
	bool operator() (const edge& a, const edge& b)
	{
		return a.value > b.value;
	}
};

int main(void) // https://www.acmicpc.net/problem/1916
{
	cin.tie(0)->sync_with_stdio(0);
	int n = 0, m = 0;
	cin >> n >> m;
	vector<vector<edge>>graph(n + 1);
	for (int i = 0; i < m; ++i)
	{
		int v1 = 0, v2 = 0;
		ll num = 0;
		cin >> v1 >> v2 >> num;
		graph[v1].push_back({ v2,num });
	}
	int start = 0, end = 0;
	cin >> start >> end;
	priority_queue<edge, vector<edge>, cmp>pq;
	vector<ll>dist(n + 1, LLONG_MAX);
	dist[start] = 0;
	pq.push({ start,0 });
	while (!pq.empty())
	{
		edge cur = pq.top();
		pq.pop();
		if (dist[cur.vertex] < cur.value)
		{
			continue;
		}
		for (int i = 0; i < (int)graph[cur.vertex].size(); ++i)
		{
			int ver = graph[cur.vertex][i].vertex;
			ll num = graph[cur.vertex][i].value;
			if (dist[ver] > dist[cur.vertex] + num)
			{
				dist[ver] = dist[cur.vertex] + num;
				pq.push({ ver,dist[ver] });
			}
		}
	}
	cout << dist[end];
	return 0;
}

File: B트랙/8주차/참고 코드/dijkstra.java
import java.io.*;
import java.util.*;

class Edge implements Comparable<Edge>
{
    int vertex;
    long value;
    public Edge(int vertex, long value)
    {
        this.vertex = vertex;
        this.value = value;
    }
    @Override
    public int compareTo(Edge other)
    {
        return Long.compare(this.value, other.value);
    }
}

public class Main
{
    public static void main(String[] args) throws IOException // https://www.acmicpc.net/problem/1916
    {
        BufferedReader scan = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(scan.readLine());
        int n = Integer.parseInt(st.nextToken());
        st = new StringTokenizer(scan.readLine());
        int m = Integer.parseInt(st.nextToken());
        List<List<Edge>> graph = new ArrayList<>();
        for (int i = 0; i <= n; i++)
        {
            graph.add(new ArrayList<>());
        }
        for (int i = 0; i < m; i++)
        {
            st = new StringTokenizer(scan.readLine());
            int v1 = Integer.parseInt(st.nextToken());
            int v2 = Integer.parseInt(st.nextToken());
            long num = Long.parseLong(st.nextToken());
            graph.get(v1).add(new Edge(v2, num));
        }
        st = new StringTokenizer(scan.readLine());
        int start = Integer.parseInt(st.nextToken());
        int end = Integer.parseInt(st.nextToken());
        long[] dist = new long[n + 1];
        Arrays.fill(dist, Long.MAX_VALUE);
        dist[start] = 0;
        PriorityQueue<Edge> pq = new PriorityQueue<>();
        pq.offer(new Edge(start, 0));
        while (!pq.isEmpty())
        {
            Edge cur = pq.poll();
            if (dist[cur.vertex] < cur.value)
            {
                continue;
            }
            for (Edge next : graph.get(cur.vertex))
            {
                if (dist[next.vertex] > dist[cur.vertex] + next.value)
                {
                    dist[next.vertex] = dist[cur.vertex] + next.value;
                    pq.offer(new Edge(next.vertex, dist[next.vertex]));
                }
            }
        }
        System.out.println(dist[end]);
        scan.close();
    }
}
